tree/ntupleutil/v7/src/RNTupleImporter.cxx

```diff
@@ -81,32 +81,54 @@ ROOT::Experimental::RNTupleImporter::RLeafArrayTransformation::Transform(const R
 }
 
 ROOT::Experimental::RResult<std::unique_ptr<ROOT::Experimental::RNTupleImporter>>
-ROOT::Experimental::RNTupleImporter::Create(std::string_view sourceFile, std::string_view treeName,
-                                            std::string_view destFile)
+ROOT::Experimental::RNTupleImporter::Create(std::string_view sourceFileName, std::string_view treeName,
+                                            std::string_view destFileName)
 {
    auto importer = std::unique_ptr<RNTupleImporter>(new RNTupleImporter());
    importer->fNTupleName = treeName;
-   importer->fSourceFile = std::unique_ptr<TFile>(TFile::Open(std::string(sourceFile).c_str()));
+   importer->fSourceFile = std::unique_ptr<TFile>(TFile::Open(std::string(sourceFileName).c_str()));
    if (!importer->fSourceFile || importer->fSourceFile->IsZombie()) {
-      return R__FAIL("cannot open source file " + std::string(sourceFile));
+      return R__FAIL("cannot open source file " + std::string(sourceFileName));
    }
    importer->fSourceTree = std::unique_ptr<TTree>(importer->fSourceFile->Get<TTree>(std::string(treeName).c_str()));
    if (!importer->fSourceTree) {
-      return R__FAIL("cannot read TTree " + std::string(treeName) + " from " + std::string(sourceFile));
+      return R__FAIL("cannot read TTree " + std::string(treeName) + " from " + std::string(sourceFileName));
    }
    // If we have IMT enabled, its best use is for parallel page compression
    importer->fSourceTree->SetImplicitMT(false);
 
-   importer->fDestFileName = destFile;
-   importer->fWriteOptions.SetCompression(kDefaultCompressionSettings);
-   importer->fDestFile = std::unique_ptr<TFile>(TFile::Open(importer->fDestFileName.c_str(), "UPDATE"));
-   if (!importer->fDestFile || importer->fDestFile->IsZombie()) {
-      return R__FAIL("cannot open dest file " + std::string(importer->fDestFileName));
-   }
+   importer->SetupDestination(destFileName);
+
+   return importer;
+}
+
+ROOT::Experimental::RResult<std::unique_ptr<ROOT::Experimental::RNTupleImporter>>
+ROOT::Experimental::RNTupleImporter::Create(TTree *sourceTree, std::string_view destFileName)
+{
+   auto importer = std::unique_ptr<RNTupleImporter>(new RNTupleImporter());
+   importer->fNTupleName = sourceTree->GetName();
+   importer->fSourceTree = std::unique_ptr<TTree>(sourceTree);
```

Comment:
That looks odd.  The interface (`Create(TTree*`) does not indicate that the callee is taking ownership of the passed object (and actually it probably should not).

Reply:
Ah yes you're right (coming from a non-C++ background the ownership semantics are still a bit new to me). What would be the best way to go about it here? The other `Create` method does own the tree object (and I think has to, since only the tree name and source file are passed).

Reply:
Even in the case where the file is opened explicitly by this code (and thus the file object is owned by the importer), there is no need to take ownership of the `TTree` object as the file shares ownership of the tree, i.e. if the tree has not already been deleted,it will delete it.   (Side note, as a consequence in the current code the order in which the 2 unique_ptr are declared is essential;  They are currently in the right order, if they had been reversed then after the deletion of the file pointer, the `TTree` pointer would have been dangling).

Reply:
Got it, so we can/should get rid of the `unique_ptr` entirely?

Reply:
For the tree yes :)